bench/rust-analyzer-rs/optimized/1yigtuxmiuvrwpzt.ll

```diff
@@ -14671,29 +14671,54 @@ _ZN12tracing_core8callsite15DefaultCallsite8interest17hccd71802e1b704feE.exit109
-  %265 = icmp eq i64 %264, 18
-  br i1 %265, label %"_ZN4core3ptr173drop_in_place$LT$lock_api..rwlock..RwLockReadGuard$LT$parking_lot..raw_rwlock..RawRwLock$C$salsa..derived..slot..QueryState$LT$hir_expand..db..ProcMacroSpanQuery$GT$$GT$$GT$17h82b672e501d209dcE.exit.sink.split.i38", label %87
+  %.val31 = load ptr, ptr %262, align 8
+  switch i32 %86, label %default.unreachable61 [
+    i32 3, label %273
+    i32 1, label %263
+    i32 2, label %268
+  ]
 
-"_ZN4core3ptr173drop_in_place$LT$lock_api..rwlock..RwLockReadGuard$LT$parking_lot..raw_rwlock..RawRwLock$C$salsa..derived..slot..QueryState$LT$hir_expand..db..ProcMacroSpanQuery$GT$$GT$$GT$17h82b672e501d209dcE.exit.sink.split.i38": ; preds = %261
+"_ZN4core3ptr173drop_in_place$LT$lock_api..rwlock..RwLockReadGuard$LT$parking_lot..raw_rwlock..RawRwLock$C$salsa..derived..slot..QueryState$LT$hir_expand..db..ProcMacroSpanQuery$GT$$GT$$GT$17h82b672e501d209dcE.exit.sink.split.i38": ; preds = %273, %268, %263
   call void @_ZN11parking_lot10raw_rwlock9RawRwLock18unlock_shared_slow17h48c5de16cc1f8475E(ptr noundef nonnull align 8 %.val31)
   br label %87
 
-266:                                              ; preds = %267, %"_ZN5salsa7derived4slot18Slot$LT$Q$C$MP$GT$12read_upgrade17h2e49ea84699dd689E.exit"
+266:                                              ; preds = %261
+  %264 = icmp ne ptr %.val31, null
+  call void @llvm.assume(i1 %264)
+  %265 = atomicrmw sub ptr %.val31, i64 16 release, align 8
+  %266 = and i64 %265, -14
+  %267 = icmp eq i64 %266, 18
+  br i1 %267, label %"_ZN4core3ptr173drop_in_place$LT$lock_api..rwlock..RwLockReadGuard$LT$parking_lot..raw_rwlock..RawRwLock$C$salsa..derived..slot..QueryState$LT$hir_expand..db..ProcMacroSpanQuery$GT$$GT$$GT$17h82b672e501d209dcE.exit.sink.split.i38", label %87
+
+268:                                              ; preds = %261
+  %269 = icmp ne ptr %.val31, null
+  call void @llvm.assume(i1 %269)
+  %270 = atomicrmw sub ptr %.val31, i64 16 release, align 8
+  %271 = and i64 %270, -14
+  %272 = icmp eq i64 %271, 18
+  br i1 %272, label %"_ZN4core3ptr173drop_in_place$LT$lock_api..rwlock..RwLockReadGuard$LT$parking_lot..raw_rwlock..RawRwLock$C$salsa..derived..slot..QueryState$LT$hir_expand..db..ProcMacroSpanQuery$GT$$GT$$GT$17h82b672e501d209dcE.exit.sink.split.i38", label %87
+
+273:                                              ; preds = %261
+  %274 = icmp ne ptr %.val31, null
+  call void @llvm.assume(i1 %274)
+  %275 = atomicrmw sub ptr %.val31, i64 16 release, align 8
+  %276 = and i64 %275, -14
+  %277 = icmp eq i64 %276, 18
+  br i1 %277, label %"_ZN4core3ptr173drop_in_place$LT$lock_api..rwlock..RwLockReadGuard$LT$parking_lot..raw_rwlock..RawRwLock$C$salsa..derived..slot..QueryState$LT$hir_expand..db..ProcMacroSpanQuery$GT$$GT$$GT$17h82b672e501d209dcE.exit.sink.split.i38", label %87
+
+278:                                              ; preds = %279, %"_ZN5salsa7derived4slot18Slot$LT$Q$C$MP$GT$12read_upgrade17h2e49ea84699dd689E.exit"
```

Comment:
Seems like regression, but the lengths of generated ASMs are approaching.

https://godbolt.org/z/8xzWW131c